### ISSUE TITLE
JSONToTable does not support backslash character

### ISSUE THREAD

**Issue by Akulla-A** [Facepunch/garrysmod-issues#5569]
### Details
Issue concerning util.JSONToTable. The function does not support backslash. The position of the backslash changes nothing. This is not an issue about malformed JSON I suppose, online JSON parsers/viewers have no issue with the JSON below.

### Steps to reproduce
Same issue, clientside or serverside. Gmod Chromium x86-64

First and second call return nil, third call return a table:
lua_run print(util.JSONToTable('["Hi\nTest"]'), util.JSONToTable('["Hi\bTest"]'), util.JSONToTable('["HiTest"]'))
lua_run print(util.JSONToTable('{"1":"Hi\nTest"}'), util.JSONToTable('{"1":"Hi\aTest"}'), util.JSONToTable('{"1":"HiTest"}'))

However, this works fine and will return a table:
lua_run print(util.JSONToTable(util.TableToJSON({"Hi\nTest"})))

So, calling JSONToTable with a TableToJSON result as a parameter works fine with any of the parameters above. 
However, calling JSONToTable with something else than a direct TableToJSON result ( Manually C/P TableToJSON result, or a SQL Query) break it.
Maybe there is "hidden" characters when printing in the console that are necessary for the JSONToTable?

**Comment by CornerPin** [comment]
The function actually has no problems with handling backslashes, the real issue you are having is that you don't properly [escape](https://en.wikipedia.org/wiki/Escape_character) backslash characters in your string literals.

**TL;DR: replace `\` with `\\` in your first two examples.**

In Lua, as in many programming languages, string literals can contain escape sequences, which are replaced with characters they represent when parsed. In your examples, your strings contain sequences [`\n`](https://en.wikipedia.org/wiki/Newline), [`\b`](https://en.wikipedia.org/wiki/Backspace) and [`\a`](https://en.wikipedia.org/wiki/Bell_character), which represent [newline](https://en.wikipedia.org/wiki/Newline), [backspace](https://en.wikipedia.org/wiki/Backspace) and [bell](https://en.wikipedia.org/wiki/Bell_character) characters, respectively. Because they are not escaped, Lua replaces them with their [control characters](https://en.wikipedia.org/wiki/Control_character#In_ASCII), e.g. string `Hi\nTest` turns into
```
Hi
Test
```
when parsed.

Quoting [section 7 of **RFC 8259 - The JavaScript Object Notation (JSON) Data Interchange Format**](https://datatracker.ietf.org/doc/html/rfc8259#section-7):

> A string begins and ends with quotation marks.  All Unicode characters may be placed within the quotation marks, except for the characters that MUST be escaped: quotation mark, reverse solidus, and the control characters (U+0000 through U+001F).

```
string = quotation-mark *char quotation-mark

char = unescaped /
    escape (
        %x22 /          ; "    quotation mark  U+0022
        %x5C /          ; \    reverse solidus U+005C
        %x2F /          ; /    solidus         U+002F
        %x62 /          ; b    backspace       U+0008
        %x66 /          ; f    form feed       U+000C
        %x6E /          ; n    line feed       U+000A
        %x72 /          ; r    carriage return U+000D
        %x74 /          ; t    tab             U+0009
        %x75 4HEXDIG )  ; uXXXX                U+XXXX

escape = %x5C              ; \

quotation-mark = %x22      ; "

unescaped = %x20-21 / %x23-5B / %x5D-10FFFF
```
Because newline, backspace and bell characters are all outside of the `unescaped` range, they must be escaped in order to be properly parsed by `util.JSONToTable`.

The reason your third example works without any problems is because `util.TableToJSON` replaces the newline character with `\\n`, which can then be properly parsed by `util.JSONToTable`.
